### ISSUE TITLE
first password protection form token is always invalid/expired

### ISSUE THREAD

**Issue by beckysoll** [statamic/v2-hub#1525]
### Actual behaviour
It seems that the first time you are taken to the form_url that has been set for a password protected page (in this case i've protected the whole site) the token is invalid. i've tested this repeatedly by clearing my browser's cookies, visiting the page, being redirected to the `form_url`, entering the password correctly, and every single time i get the error "Invalid or expired token". I then go back to a protected page's url, get directed to the form_url again, and that time the token is accepted.

### Steps to reproduce
1.  protect the entire site with a password in `system.yaml`
2.  open a new browser, or clear your browser's cookies, visit any page in the site.
3. when you are redirected to the page that has been set as the `form_url`, enter the password
4. you will receive the error "Invalid or expired token"
5. hit back to go to the page you attempted to view
6. refresh the page
7. when you are redirected to the page that has been set as the `form_url`, enter the password
8. This time the token will be accepted 

This happens both locally in development (Via mamp) and on our server.

### Server Details

**Operating System**: mac os 10.12.5

**Web Server:** apache

**PHP Version:** 5.6.21 (server) and 7.0.8 (mamp)

**Statamic Version:** 2.6 and 2.6.1 (an issue both before and after i upgraded to the latest)

**Updated from an older Statamic or fresh install:** updated

### Logs
nothing in the logs

**Comment by jasonvarga** [comment]
Looks like a bug to me!

I find that it happens when you have a fresh session.

ie. If you already have the site open, _then_ apply the sitewide protection, your password will work first try.
But if you _haven't_ had the site open, and then try to enter the password, you'll see this behavior.

I must have missed it because I was toggling the setting on and off without closing out my session.

Good find.

**Comment by beckysoll** [comment]
Yeah, that sounds about right. Thanks, just wanted to make sure it wasn't me!

**Comment by gmcz** [comment]
So, I seem to be getting this error on any page which I set to password protection. Any fix in the works? A patch I can apply real quick?

**Comment by gmcz** [comment]
As a temporary work-around I'm making folks land on page 1, then click a link to page 2 (which is the password protected one). Not elegant, but working for now. 

If I'm going to try and write a patch, would I want to create a session prior to the redirect exception that happens in PasswordProtector.php, for example?

**Comment by beckysoll** [comment]
I hope a fix is coming, it’s definitely still an issue for me.

**Comment by edalzell** [comment]
I'm curious why there's no session...why isn't it created?

**Comment by gmcz** [comment]
Seems to me, @edalzell, that the redirect exception happens before a session is created. Thus the visitor ends up on the password entry url with no session cookies set.

**Comment by jasonvarga** [comment]
@beckysoll @gmcz Want to throw this in `statamic/core/Http/Controllers/StatamicController.php` and see if it works for you?

![image](https://user-images.githubusercontent.com/105211/37526666-5d2ce9d2-2906-11e8-837a-4399ee864e21.png)

Here it is in copy-pastable format:

```
        try {
            $this->protect();
        } catch (\Statamic\Exceptions\RedirectException $e) {
            return redirect($e->getUrl(), $e->getCode());
        }
```

To most easily replicate the bug and test the fix:

- Open an incognito window and hit your site. Bug should be there.
- Close window
- Apply patch
- Open another incognito window and hit again. Bug should be gone.🤞

**Comment by beckysoll** [comment]
It worked, woohoo!

**Comment by gmcz** [comment]
Perfect! Thank you ever so much @jasonvarga!
